File: batch-tech-feed/produce/techfeed/renderer.py
"""Remotion を呼び出して mp4 を焼く。

映像の合成は Remotion に任せる。音声も Remotion 側で多重化するため、ここでは
manifest を props として渡すだけでよい(docs/407 §2.4)。
"""

from __future__ import annotations

import json
import shutil
import subprocess
from pathlib import Path

# remotion プロジェクトの場所。batch-tech-feed/remotion/ に置いてある。
PROJECT_DIR = Path(__file__).resolve().parent.parent / "remotion"
ENTRY = "src/index.ts"
COMPOSITION = "Digest"


class RenderError(RuntimeError):
    pass


def run(
    manifest: dict,
    out_dir: str,
    dst: str,
    concurrency: int | None = None,
    crf: int = 20,
    quiet: bool = False,
) -> str:
    if not (PROJECT_DIR / "node_modules").is_dir():
        raise RenderError(
            f"{PROJECT_DIR}/node_modules not found; run `npm install` in that directory first"
        )
    if shutil.which("npx") is None:
        raise RenderError("npx not found on PATH (Node.js is required to render)")

    # props はファイル渡しにする。manifest は画像を data URI で抱えるため、
    # コマンドライン引数に載せると容易に長さ制限を超える。
    #
    # 中身は必ず Composition の props の形 ({"manifest": ...}) にすること。manifest を
    # 直に書くと defaultProps とマージされて既定値が生き残り、尺が既定のまま
    # 静かに短い動画が出てくる(エラーにならないので気づきにくい)。
    props_path = Path(out_dir) / "props.json"
    props_path.write_text(
        json.dumps({"manifest": manifest}, ensure_ascii=False), encoding="utf-8"
    )

    cmd = [
        "npx", "remotion", "render", ENTRY, COMPOSITION, str(Path(dst).resolve()),
        f"--props={props_path.resolve()}",
        # ナレーション wav をブラウザから読ませるため、作業ディレクトリを public 扱いにする。
        # 画像は manifest 内で data URI に畳んであるので、ここに置く必要があるのは音声だけ。
        f"--public-dir={Path(out_dir).resolve()}",
        f"--crf={crf}",
    ]
    if concurrency:
        cmd.append(f"--concurrency={concurrency}")
    if quiet:
        cmd.append("--log=error")

    proc = subprocess.run(cmd, cwd=PROJECT_DIR)
    if proc.returncode != 0:
        raise RenderError(f"remotion render failed with exit code {proc.returncode}")
    return dst

File: batch-tech-feed/produce/shared/renderer.go
package shared

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
)

const (
	RemotionEntry       = "src/index.ts"
	RemotionComposition = "Digest"
)

func RunRender(manifest *Manifest, outDir, dst string, concurrency, crf int, quiet bool) (string, error) {
	ex, err := os.Executable()
	var baseDir string
	if err == nil {
		baseDir = filepath.Dir(ex)
	}
	projectDir, err := filepath.Abs(filepath.Join(baseDir, "remotion"))
	if err != nil || !isDir(projectDir) {
		projectDir, _ = filepath.Abs("remotion")
	}

	if !isDir(filepath.Join(projectDir, "node_modules")) {
		return "", fmt.Errorf("%s/node_modules not found; run `npm install` in that directory first", projectDir)
	}

	npxPath, err := exec.LookPath("npx")
	if err != nil {
		return "", fmt.Errorf("npx not found on PATH (Node.js is required to render)")
	}

	propsPath := filepath.Join(outDir, "props.json")
	propsMap := map[string]interface{}{
		"manifest": manifest,
	}
	propsBytes, err := json.Marshal(propsMap)
	if err != nil {
		return "", fmt.Errorf("failed to marshal props JSON: %w", err)
	}
	if err := os.WriteFile(propsPath, propsBytes, 0o644); err != nil {
		return "", fmt.Errorf("failed to write props file: %w", err)
	}

	absDst, err := filepath.Abs(dst)
	if err != nil {
		absDst = dst
	}

	absProps, err := filepath.Abs(propsPath)
	if err != nil {
		absProps = propsPath
	}

	absOutDir, err := filepath.Abs(outDir)
	if err != nil {
		absOutDir = outDir
	}

	args := []string{
		"remotion", "render", RemotionEntry, RemotionComposition, absDst,
		fmt.Sprintf("--props=%s", absProps),
		fmt.Sprintf("--public-dir=%s", absOutDir),
		fmt.Sprintf("--crf=%d", crf),
	}

	if concurrency > 0 {
		args = append(args, fmt.Sprintf("--concurrency=%d", concurrency))
	}
	if quiet {
		args = append(args, "--log=error")
	}

	cmd := exec.Command(npxPath, args...)
	cmd.Dir = projectDir
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if !quiet {
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
	}

	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("remotion render failed (%w): %s", err, stderr.String())
	}

	return absDst, nil
}

func isDir(path string) bool {
	fi, err := os.Stat(path)
	return err == nil && fi.IsDir()
}
